src/blockchain/rpc.rs

```diff
@@ -0,0 +1,522 @@
+    capabilities: HashSet<Capability>,
+    rpc_version: usize,
+    wallet_name: String,
+
+    // This is a fixed Address used as a key to store information on the node
+    satoshi_address: Address,
+}
+
+#[derive(Debug)]
+pub struct RpcConfig {
+    pub url: String,
+    pub auth: Auth,
+    pub network: Network,
+
+    pub descriptor: ExtendedDescriptor,
+    pub change_descriptor: Option<ExtendedDescriptor>,
+}
+
+impl RpcBlockchain {
+    fn get_node_synced_height(&self) -> Result<u32, Error> {
+        let info = self.client.get_address_info(&self.satoshi_address)?;
+        if let Some(GetAddressInfoResultLabel::Simple(label)) = info.labels.first() {
+            Ok(label.parse::<u32>().unwrap_or(0))
+        } else {
+            Ok(0)
+        }
+    }
+
+    fn set_node_synced_height(&self, height: u32) -> Result<(), Error> {
+        Ok(self
+            .client
+            .set_label(&self.satoshi_address, &height.to_string())?)
+    }
+}
+
+impl Blockchain for RpcBlockchain {
+    fn get_capabilities(&self) -> HashSet<Capability> {
+        self.capabilities.clone()
+    }
+
+    fn setup<D: BatchDatabase, P: 'static + Progress>(
+        &self,
+        stop_gap: Option<usize>,
+        database: &mut D,
+        progress_update: P,
+    ) -> Result<(), Error> {
+        let mut scripts_pubkeys = database.iter_script_pubkeys(Some(KeychainKind::External))?;
+        scripts_pubkeys.extend(database.iter_script_pubkeys(Some(KeychainKind::Internal))?);
+        debug!(
+            "importing {} script_pubkeys (some maybe already imported)",
+            scripts_pubkeys.len()
+        );
+        let requests: Vec<_> = scripts_pubkeys
+            .iter()
+            .map(|s| ImportMultiRequest {
+                timestamp: ImportMultiRescanSince::Timestamp(0),
+                script_pubkey: Some(ImportMultiRequestScriptPubkey::Script(&s)),
+                watchonly: Some(true),
+                ..Default::default()
+            })
+            .collect();
+        let options = ImportMultiOptions {
+            rescan: Some(false),
+        };
+        // Note we use import_multi because as of bitcoin core 0.21.0 many descriptors are not supported
+        // https://bitcoindevkit.org/descriptors/#compatibility-matrix
+        //TODO maybe convenient using import_descriptor for compatible descriptor and import_multi as fallback
+        self.client.import_multi(&requests, Some(&options))?;
+        self.sync(stop_gap, database, progress_update)
+    }
+
+    fn sync<D: BatchDatabase, P: 'static + Progress>(
+        &self,
+        _stop_gap: Option<usize>,
+        db: &mut D,
+        progress_update: P,
+    ) -> Result<(), Error> {
+        debug!("sync");
+        let current_height = self.get_height()?;
+        let node_synced = self.get_node_synced_height()?;
+
+        //TODO if current_height == node_synced should check only the mempool
+
+        //TODO split the interval in chunk so that we can give progress update
```

Comment:
instead of start/stop params which involves state, we could just periodically call `getwalletinfo` for the progress.

Reply:
yes, that is better

Reply:
reconsidered again, see https://github.com/bitcoindevkit/bdk/pull/348#discussion_r636016580